src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,528 @@
+			}
+
+			/// <summary>
+			/// Unlock the context, causing <see cref="Wait"/> to cease blocking. Do nothing if the context
+			/// is not locked.
+			/// </summary>
+			protected virtual void Unlock()
+			{
+				// Do not set _locked back to false, we want to wait only once per transaction context.
+				// Volatile member, do not simplify by removing the local variable.
+				var semaphore = _semaphore;
+				_semaphore = null;
+				if (semaphore == null)
+					return;
+				// Not supposed to have more than one waiting thread, but in case the user made bad multi-threading,
+				// do not leave frozen threads.
+				semaphore.Release(100);
+				semaphore.Dispose();
+			}
+
+			#region IEnlistmentNotification Members
+
+			/// <summary>
+			/// Prepare the session for the transaction commit. Runs
+			/// <see cref="ISessionImplementor.BeforeTransactionCompletion(ITransaction)"/> for the session and for
+			/// <see cref="ConnectionManager.DependentSessions"/> if any. <see cref="Lock"/> the context
+			/// before signaling it is done, or before rollback in case of failure.
+			/// </summary>
+			/// <param name="preparingEnlistment">The object for notifying the prepare phase outcome.</param>
+			public virtual void Prepare(PreparingEnlistment preparingEnlistment)
+			{
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+				{
+					try
+					{
+						using (_sessionImplementor.ConnectionManager.BeginsProcessingFromSystemTransaction(_useConnectionOnSystemTransactionEvents))
+						{
+							if (_useConnectionOnSystemTransactionEvents)
+							{
+								// Ensure any newly acquired connection gets enlisted in the transaction. When distributed,
+								// this code runs from another thread and we cannot rely on Transaction.Current.
+								using (var tx = new TransactionScope(AmbientTransaction))
+								{
+									_sessionImplementor.BeforeTransactionCompletion(null);
+									foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+										dependentSession.BeforeTransactionCompletion(null);
+
+									tx.Complete();
+								}
+							}
+							else
+							{
+								_sessionImplementor.BeforeTransactionCompletion(null);
+								foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+									dependentSession.BeforeTransactionCompletion(null);
+							}
+						}
+						// Lock the session to ensure second phase gets done before the session is used by code following
+						// the transaction scope disposal.
+						Lock();
+
+						_logger.Debug("Prepared and done for system transaction");
+						// We do not have anything more to do in second phases callbacks: the remaining work is handled in
+						// transaction completion event, which always executes, contrary to second phases callbacks which
+						// do not, depending on the rollback cases.
+						// This saves a thread when distributed.
+						preparingEnlistment.Done();
```

Comment:
>Volatile resource callbacks are called depending of a number of conditions, which I am not sure to have count them all, despite the ResourceManagerFixture I have added for checking that.

They are called just fine if implemented properly.

>Especially, there are cases where none of the second phase callbacks are called.

Can you show one of such cases?

>Maybe we can "have it right" without the TransactionCompleted event, but I think this will require more complex code. 

The code would be much simpler, because there will be no try-catch on getting `wasSuccessful`, there will be no unsubscription from the delegate, etc.

>About "another volatile resource", why?

Because this would behave the same as it done right now.

>Resources are not sequentially called. 

They are. It's just a for loop.  But it does not really matter as in this case we would have one resource with empty prepare and one with empty completion.